lib/policy_machine_storage_adapters/active_record.rb

```diff
@@ -772,6 +772,65 @@ def accessible_objects(user_or_attribute, operation, options = {})
       end
     end
 
+    # Version of accessible_objects which only returns objects that are
+    # ancestors of a specified root object or the object itself
+    def accessible_ancestor_objects(user_or_attribute, operation, root_object, options = {})
+      # Drop down to the root's stored policy element if not already there
+      root_object_pe = root_object.respond_to?(:stored_pe) ? root_object.stored_pe : root_object
+
+      # The final set of accessible objects must be contained in the following set
+      ancestor_objects = root_object_pe.ancestors(type: class_for_type('object')) + [root_object_pe]
+
+      # Short-circuit and return all ancestors (minus prohibitions), if authorized on the root node
+      if is_privilege?(user_or_attribute, operation, root_object_pe)
```

Comment:
How about extracting parts of this block into a method(s)? It seems like there's a lot of similarity with the functionality in different parts of the logic tree.